source/mainnet/smart-contracts/tutorials/voting/voting-sc.rst

```diff
@@ -86,4 +85,20 @@ And if the end time has passed, an error occurs.
 Initializing
 ------------
 
-The election is open from the point in time that this smart contract is initialized until the `end_time`.
+The election is open from the point in time that this smart contract is initialized until the deadline.
+
+Performance considerations
+--------------------------
+
+An important aspect of the voting smart contract to highlight is the way that votes are tallied. Note how two ``HashMap``s are used: one for which accounts have voted for which option, and one for how many votes each option has.
+However, if you think about it, only the first ``HashMap``, the ``ballots`` field is necessary. The vote counts can always be determined from those. So why do it in this way?
+
+The answer is that by counting the votes separately, we avoid having to iterate through all the ballots in order to count the votes.
+Unbounded iterations in smart contracts is in general a bad idea. For instance, if there are many votes, executing the smart contract might become costly, as iterating through the ballots takes many operations.
+In the worst case, this may exceed the energy budget of running the functions, which could make the smart contract unusable.
+
+It's important to keep these performance considerations in mind when writing smart contracts. In general, writing smart contracts requires a different methodology than most other usual software development.
+Be sure to think carefully and read up on the common issues and security problems that may occur when you develop smart contracts.
+
+
+In the next part of the tutorial, we will set up a frontend to make it easier to interact with the smart contract.
```

Comment:
A few things to improve/consider:

- Rhetoric questions might not be needed, you can explain the reason for the architectural design in a concise manner without it.

---------------------------------------------
Your description in the smart contract seems quite clear, I suggest re-using it here:
```
    /// Note the data duplication here! The tally can of course always be
    /// determined by examining the ballots field. This requires looping
    /// over all ballots, counting how many votes exist for a specific option.
    /// However, such a loop would be practically unbounded, as it is only
    /// limited by the number of votes. Such a loop could exhaust the energy
    /// budget of the smart contract transaction, potentially making the smart
    /// contract unusable and vulnerable to a kind of DDoS attack.
    ///
    /// Thus, we favor duplicating the data in another hashmap, where the vote
    /// count can be retrieved and updated in constant time. 
```



In the worst case, this may exceed the energy budget of the ``transaction``, which could make the smart contract unusable.


People could ask why not just increase the `energy budget of the transaction` (this is a value set by users and can be set to a very high value).

There are limits either by e.g. the block gas/energy limit or by the limit/funds that the account has at its disposal (you can spend in gas/energy only as much as you have funds on the account). Not sure how to communicate this concisely so.

Reply:
I reworded/used what I had already written in the code :)